[AMDGPU] NFC: Provide RPTracker interface for external iterators

jrbyrnes opened the pull request:
This is part of a series of PRs which enable using the GCNRPTrackers during scheduling. I've split them up to (hopefully) make reviewing easier. For context see https://github.com/llvm/llvm-project/pull/88797 . 

This PR adds adds an interface to the existing GCNRPTrackers which provides a way to use the trackers with an externally managed iterator.

jrbyrnes committed:
[AMDGPU] NFC: Provide RPTracker interface for external iterators

Change-Id: I79b54722e6e858961486248d94766c3f3c161160

arsenm commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -343,24 +343,25 @@ void GCNRPTracker::reset(const MachineInstr &MI,
+////////////////////////////////////////////////////////////////////////////////
+// GCNUpwardRPTracker
+
+bool GCNUpwardRPTracker::recede(const MachineInstr &MI, bool ShouldTrackIt) {
```
Doxygen comment for what this does? And the parameters and return? Maybe should go on the header, instead of the definition

arsenm commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -430,28 +432,44 @@ bool GCNDownwardRPTracker::reset(const MachineInstr &MI,
+             ? CurrLIS->getInstructionIndex(*LastTrackedMI).getDeadSlot()
+             : CurrLIS->getInstructionIndex(*NextMI).getBaseIndex();
+  }
 
-  assert(NextMI == MBBEnd || !NextMI->isDebugInstr());
+  else { //! ShouldTrackIt
```
else on same line as closing }

jrbyrnes committed:
Review comments

Change-Id: Ib798a9d6add7d6dd1ccd0e01bea09ac2f4eeb94f

vpykhtin commented:
I haven't found cases where _advance_ or _recede_ arecalled with ShouldTrackIt = false, and the semantics of the flag looks vague, can we avoid the flag?

jrbyrnes committed:
Fix RP calculation behavior

Change-Id: I10242186f538359ff09110dd70b23e5136655849

jrbyrnes commented:
> I haven't found cases where advance or recede arecalled with ShouldTrackIt = false, and the semantics of the flag looks vague, can we avoid the flag?

Thanks for catching that, that was my mistake. We should be using the !ShouldTrackIt version during scheduling. I've corrected that in https://github.com/llvm/llvm-project/pull/93090 , and also corrected an issue with the way RP was calculated using the !ShouldTrackIt version.

I hope it is more clear why ShouldTrackIt is needed in the latest version. We cannot perform speculative RP queries (initCandidate) using the internal iterator, so we need a tracker which allows us to provide the NextMI. If we did not want to introduce a flag, we would need to implement a new tracker that doesn't use an internal iterator, as was done in https://github.com/llvm/llvm-project/pull/88797

jrbyrnes committed:
Review comments

Change-Id: I5c25a2e899c46fc3026b215010bde32b6c6d24ac

jrbyrnes committed:
Remove flag from upward RPTracker

Change-Id: I9633e160b23360da06520b63a0e9b7c0fd33647a

jrbyrnes committed:
Fix bumpDeadDefs port issues

Change-Id: I1d557aa00cfb57f7a0395131ef57ac19efb5d5cf

jrbyrnes committed:
Fix bumpUpwardPressure port issues

hange-Id: Ib383230c9fc88fb59a4d3c44f4b1459050554aa5
Change-Id: I0333ac7f2000e9c8fccb369e330882530ca0c912

jrbyrnes committed:
Port getRegLanes

Change-Id: I1cc390ff9a10fac9e874b097e3279376e52209a3

jrbyrnes committed:
Port getLanesWithProperty

Change-Id: Idba7806fd4d6d43ba0381d7731720ac358c108e6

jrbyrnes committed:
Port getLastUsedLanes

Change-Id: I233f0d3b3ce89adbc5a5c22efa2f28d5fb438e4a

jrbyrnes committed:
Fix port issues

Change-Id: I78f62d348b90448fa934aee1e9900634bf8dc209

jrbyrnes committed:
Port bumpDeadDefs

Change-Id: I3b23fd089a1350570162633a4318feca092e472a

jrbyrnes committed:
Port bumpUpwardPressure

Change-Id: Ie030633023361be626a34a1867e7740777a575d9

jrbyrnes committed:
Port bumpDownwardPressure

Change-Id: Ib207ef130a024b3fd12f4c34e80f59dd23450ed2

jrbyrnes committed:
Modify behavior to support RP subreg aware RP speculation

Change-Id: I2e3c08b4487262cb290cd070f7c2042156217fb1

jrbyrnes committed:
Fix bumpDownwardPressure port issues

Change-Id: I2e3c08b4487262cb290cd070f7c2042156217fb1

jrbyrnes committed:
Port findUseBetween

Change-Id: I523e493fc80247497fd53925f05b9a8ad5f3771a

jrbyrnes commented:
In the latest, I've ported over RP speculation code from the generic trackers to GCN Trackers. While this does introduce code duplication, I thought this to be the better approach than to try to support subreg liveness in generic speculation and introduce generic tracker copies per speculation.

I've tried to organize the git history to ease the burden of review, please let me know if there's something I should do to assist further.

Commits named "Port *" are simple copy/paste of code./
Commits named "Fix * port issues" are minor changes to resolve API / build errors

The most interesting is "Modify behavior to support RP subreg aware RP speculation" which slightly modifies the behavior to support accurate RP speculation.

arsenm commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -288,6 +288,72 @@ collectVirtualRegUses(SmallVectorImpl<RegisterMaskPair> &RegMaskPairs,
+    }
+
+    return Result;
+  } else {
```
No else after return

arsenm commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -288,6 +288,72 @@ collectVirtualRegUses(SmallVectorImpl<RegisterMaskPair> &RegMaskPairs,
+getLanesWithProperty(const LiveIntervals &LIS, const MachineRegisterInfo &MRI,
+                     bool TrackLaneMasks, Register RegUnit, SlotIndex Pos,
+                     LaneBitmask SafeDefault,
+                     bool (*Property)(const LiveRange &LR, SlotIndex Pos)) {
```
can you use a function_ref here

arsenm commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -288,6 +288,72 @@ collectVirtualRegUses(SmallVectorImpl<RegisterMaskPair> &RegMaskPairs,
+                                  const MachineRegisterInfo &MRI,
+                                  const LiveIntervals *LIS,
+                                  bool Upward = false) {
+  const TargetRegisterInfo &TRI = *MRI.getTargetRegisterInfo();
```
Use SIRegisterInfo

arsenm commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -288,6 +288,72 @@ collectVirtualRegUses(SmallVectorImpl<RegisterMaskPair> &RegMaskPairs,
+                                  const LiveIntervals *LIS,
+                                  bool Upward = false) {
+  const TargetRegisterInfo &TRI = *MRI.getTargetRegisterInfo();
+  for (const MachineOperand &MO : MRI.use_nodbg_operands(Reg)) {
+    if (MO.isUndef())
+      continue;
```
This probably doesn't handle subregister defs that read the register correctly. I think you need to one of the other iterators, and MO.readsReg

jrbyrnes replied:
Actually, I'm not quite sure what we're trying to accomplish here?

```
  // Definitions

  1.  undef %63.sub1:vreg_128 = COPY %49.sub18:areg_1024 
  2.  %49.sub1:areg_1024 = COPY %63.sub1:vreg_128
  3.  undef %63.sub0:vreg_128 = COPY %49.sub19:areg_1024

  // No uses of %49.sub19 or defs of %49.subx
```

If we were speculating the RP from schedule (1, 3, ..)  we would invoke findUseBetween to find any uses of %49 between 3s original position and its candidate position. I think you are saying that we should consider the %49.sub1 def in 2. as an implicit use of %49.sub19? I know that the LIS requires subregister liveness to be modeled in this way somewhat (e.g. no multiple connected components), but from the perspective of RegisterPressure, It seems more accurate (and safe) to not model it this way -- that is, to only check use operands.

arsenm replied:
It is a use, and I don't see how it could possibly be safer to ignore it as such. Deviation from the liveness model requires a much stronger rationale than this. I've fixed too many bugs from missing uses on subreg defs

jrbyrnes replied:
Yes you're right.

Ultimately the RP should agree with interference in RA, which is based on LIS subreg modeling.

That said, this seems like a shortcoming of subreg liveness modelling.

jrbyrnes replied:
Ultimately, RP should agree with RA interference. AFAICT, RA interference is determined by the subrange segments, rather than the main range with the strange liveness concerns -- I'm still analyzing.

In https://godbolt.org/z/MdTEG44a4 we have a use `%49.sub0:areg_1024` occuring before a def of the same superreg `%49.sub1:areg_1024`. According to the subreg def is a use of other subregs, `%49.sub0:areg_1024` should be live until the def `%49.sub1:areg_1024`. However, when assigning `%2:agpr_32`, RA sees there is no interference with `%49.sub0:areg_1024` and assigns to the same PhysReg.

This is consistent with how the GCNTrackers currently calculate RP: we consider the use subrange and whether or not it is live at some given index https://github.com/llvm/llvm-project/blob/10407be542aeb2b59477b167bbba3716538dc722/llvm/lib/Target/AMDGPU/GCNRegPressure.cpp#L458 and don't check if there is some reaching subreg def.

jrbyrnes committed:
Review comments

Change-Id: I2de464b32d3c6ed9a77cbbc669d735dde63c2e47

arsenm commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -317,14 +317,14 @@ getLanesWithProperty(const LiveIntervals &LIS, const MachineRegisterInfo &MRI,
   }
+
+  const LiveRange *LR = LIS.getCachedRegUnit(RegUnit);
+  // Be prepared for missing liveranges: We usually do not compute liveranges
+  // for physical registers on targets with many registers (GPUs).
```
This doesn't have anything to do with many registers or GPUs

arsenm commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -333,19 +333,21 @@ getLanesWithProperty(const LiveIntervals &LIS, const MachineRegisterInfo &MRI,
+  for (const MachineOperand &MO : MRI.reg_nodbg_operands(Reg)) {
     if (MO.isUndef())
       continue;
+    if (!MO.readsReg())
```
It's this instead of isUndef, not with

arsenm commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -517,7 +519,9 @@ void GCNUpwardRPTracker::bumpUpwardPressure(const MachineInstr *MI) {
-  const TargetRegisterInfo *TRI = MRI->getTargetRegisterInfo();
+
+  const SIRegisterInfo *TRI =
+      MI->getMF()->getSubtarget<GCNSubtarget>().getRegisterInfo();
```
Shouldn't need to crawl through MI to get o the function? This should be locally available somewhere?

arsenm commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -733,7 +737,8 @@ void GCNDownwardRPTracker::bumpDownwardPressure(const MachineInstr *MI) {
   RegisterOperands RegOpers;
-  const TargetRegisterInfo *TRI = MRI->getTargetRegisterInfo();
+  const SIRegisterInfo *TRI =
+      MI->getMF()->getSubtarget<GCNSubtarget>().getRegisterInfo();
```
Ditto

jrbyrnes committed:
Review comments + revert to use_nodbg_operands

jrbyrnes committed:
Review comments

Change-Id: I286c9ed1ae91a68da881c6fa27f5f391102d0a9c

arsenm commented:
> In the latest, I've ported over RP speculation code from the generic trackers to GCN Trackers. While this does introduce code duplication, I thought this to be the better approach than to try to support subreg liveness in generic speculation and introduce generic tracker copies per speculation.

The generic tracker really just should support subregister liveness

arsenm commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -430,28 +593,44 @@ bool GCNDownwardRPTracker::reset(const MachineInstr &MI,
+      return NextMI == MBBEnd;
+
+    assert(NextMI == MBBEnd || !NextMI->isDebugInstr());
+    CurrLIS = const_cast<LiveIntervals *>(&LIS);
+    CurrMI = const_cast<MachineInstr *>(LastTrackedMI);
```
Can you avoid const_cast

jrbyrnes committed:
Avoid const_cast

Change-Id: I1446390178f8ed8705b0542e60a5f9f1649d8299

jrbyrnes commented:
> > In the latest, I've ported over RP speculation code from the generic trackers to GCN Trackers. While this does introduce code duplication, I thought this to be the better approach than to try to support subreg liveness in generic speculation and introduce generic tracker copies per speculation.
> 
> The generic tracker really just should support subregister liveness

That is most likely the next step of this work. 

In order to enable the GCNTrackers by default, in order to save compile time, we would need to modify the scheduler to allow for a mode whcih enables RP tracking but disables the generic RPTracker. 

I think it probably makes more sense to just improve the accuracy of the generic trackers as you suggest. The purpose of this PR then is to provide access to the improvements while the longer term solution is being worked out.

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -414,6 +534,49 @@ void GCNUpwardRPTracker::recede(const MachineInstr &MI) {
+    LaneBitmask DefLanes = P.LaneMask;
+    LaneBitmask LiveBefore = (LiveAfter & ~DefLanes) | UseLanes;
+
+    CurPressure.inc(Reg, LiveAfter, LiveAfter & LiveBefore, *MRI);
```
see https://github.com/llvm/llvm-project/pull/93090/files/80e066cd1da82d1d55b74d2b3cab62d8934d9947#r1774010200

jrbyrnes commented:
Landed with https://github.com/llvm/llvm-project/commit/17bc959961aa5792821516b547100316fc886ab4
